Compression.md

```diff
@@ -0,0 +1,145 @@
+ 
+ With this setup, all the inbound data first passes through NIO's `WebSocketDecoder` where the WebSocket frames are built. It then moves into the `PermessageDeflateDecompressor` where multiple frames comprising a message are accumulated and decompressed using `zlib`'s inflater. Subsequently, the decompressed messages are moved to the `WebSocketConnection` handler.
+
+Outbound WebSocket frames first reach the `PermessageDeflateCompressor` which compresses the data held within them and relays them to the `WebSocketEncoder`. Here the frames are marshalled into raw bytes to be written to the wire after encryption.
+
+#### 3.1 Compressor implementation
+The compressor is called `PermessageDeflateCompressor`. It is a [ChannelOutboundHandler](https://apple.github.io/swift-nio/docs/current/NIO/Protocols/ChannelOutboundHandler.html).
+
+`ChannelOutboundHandler`'s `write(context:data:promise)` method implemented here gets invoked when the previous outbound handler (`WebSocketConnection`) writes data to the channel. Here, only data frames and continuation frames are processed. A WebSocket message is either available in a single data frame or a data frame followed by sequence of continuation frames. 
+
+The compressor makes sure that we have all the data pertaining to a message accumulated. Subsequently, `zlib`'s deflater is invoked and deflated data is packed into a new WebSocketFrame, which is passed to the `WebSocketFrameEncoder`.
+
+#### 3.2 Decompressor implementation
+The decompressor is, functionally, a mirror image of the compressor. It is called `PermessageDeflateDecompressor` and is [ChannelInboundHandler](https://apple.github.io/swift-nio/docs/current/NIO/Protocols/ChannelInboundHandler.html).
+
+`ChannelInboundHandler`'s `channelRead(context:data)` method implemented here is invoked whenever a new `WebSocketFrame`s is produced by the `WebSocketDecoder`. Similar to the compressor, the decompressor processes data and continuation frames only. All the data pertaining to a message, possibly spread across continuation frames, is accumulated and `zlib`'s inflater is invoked. The inflated data is then packed into a new `WebSocketFrame` and moved into the next handler in the pipeline.
+
+#### 3.3 Configuring the Compressor and Decompressor
+[RFC7692](https://tools.ietf.org/html/rfc7692) defines four configuration options. They are actually two pairs of option, one for the client and server each:
```

Comment:
Can we change this to "They are actually two pairs of options, one each for the client and the server:"